xapian-core/include/xapian/cluster.h

```diff
@@ -0,0 +1,199 @@
+class XAPIAN_VISIBILITY_DEFAULT Similarity {
+  public:
+	virtual ~Similarity();
+	virtual double similarity(TermListGroup tlg, TermIterator a_begin, TermIterator a_end, TermIterator b_begin, TermIterator b_end) = 0;
+	virtual std::string get_description() = 0;
+};
+
+// Class for caluclating the euclidian distance between two documents
+class XAPIAN_VISIBILITY_DEFAULT EuclidianDistance : public Similarity {
+  public:
+	double similarity(TermListGroup tlg, TermIterator a_begin, TermIterator a_end, TermIterator b_begin, TermIterator b_end);
+	std::string get_description();
+};
+
+class ClusterSetIterator;
+
+// Class for storing the results returned by the Clusterer
+class XAPIAN_VISIBILITY_DEFAULT ClusterSet {
+	friend class ClusterSetIterator;
+  public:
+	// A map storing the clusterid and its corresponding Document cluster
+	std::map<clusterid, std::vector<Document> > clusters;
+	// Returns the number of clusters
+	Xapian::doccount num_of_clusters();
+	// Returns the size of a cluster with clusterid 'cid'
+	Xapian::doccount cluster_size(clusterid cid);
+	/** Returns an iterator to the start of the clusters to analyze each
+	 *  cluster individually
+	 */	
+	ClusterSetIterator begin();
+	// Returns an iterator to the end of the clusters
+	ClusterSetIterator end();
+	// Used to check the cluster at index 'i'
+	ClusterSetIterator operator[](Xapian::doccount i);
+};
+
+// A class used to iterate through the ClusterSet
+class XAPIAN_VISIBILITY_DEFAULT ClusterSetIterator {
+	friend class ClusterSet;
+  public:
+	ClusterSet cset;
+	Xapian::doccount index;
+	
+	ClusterSetIterator(ClusterSet &cset_, Xapian::doccount index_)
+	: cset(cset_), index(index_) {}
+
+	clusterid operator*() const;
+	
+	// Returns the id of the current cluster
+	clusterid get_clusterid() {
+		return index;
+	}
+
+	//DocumentSet get_clusterdocs();
+		
+	// Prefix increment operator for this iterator
+	ClusterSetIterator & operator++() {
+		index++;
+		return *this;
+	}
+	
+	// Postfix increment operator for this iterator
+	ClusterSetIterator operator++(int) {
+		ClusterSetIterator c = *this;
+		index++;
+		return c;
+	}
+
+	// Prefix decrement operator for this iterator
+	ClusterSetIterator & operator--() {
+		index--;
+		return *this;
+	}
+
+	// Postfix decrement operator for this iterator
+	ClusterSetIterator operator--(int) {
+		ClusterSetIterator c = *this;
+		index--;
+		return *this;
+	}
+};
+
+// Equality test for ClusterSetIterator 
+inline bool operator==(const ClusterSetIterator &a, const ClusterSetIterator &b) {
+	return (a.index == b.index);
+}
+
+// Inequality test for ClusterSetIterator
+inline bool operator!=(const ClusterSetIterator &a, const ClusterSetIterator &b) {
+	return !(a == b);
+}
+
+class XAPIAN_VISIBILITY_DEFAULT Clusterer {
+	//Core clustering functionality
+	virtual ClusterSet do_cluster(std::vector<docid> &docs, TermListGroup &tlg);
+};
+
+class XAPIAN_VISIBILITY_DEFAULT KMeans : public Clusterer {
```

Comment:
You aren't implementing `KMeans` in this commit (or even this PR), so I don't think this should be here at all. (Same goes for the base class above, since it's impossible to tell if this is the right API design until we're looking at code that exercises it.)